File: src/lib/correction-service.ts
/**
 * NigWrite - Plagiarism Correction Engine
 * Created by: Wabi The Tech Nurse
 *
 * Takes flagged plagiarized text and rewrites it using an LLM to produce
 * authentic, natural-sounding academic text that passes plagiarism checks.
 * Automatically re-runs plagiarism detection after rewriting to verify improvement.
 */

import ZAI from 'z-ai-web-dev-sdk';

export interface CorrectionRequest {
  flaggedText: string;
  context?: string;
  documentTitle?: string;
}

export interface CorrectionResult {
  originalText: string;
  rewrittenText: string;
  originalScore: number;
  newScore: number;
  improvement: number;
  status: 'success' | 'partial' | 'failed';
  message: string;
}

export class CorrectionService {
  private zai: ZAI | null = null;

  private async getAI(): Promise<ZAI> {
    if (!this.zai) {
      this.zai = await ZAI.create();
    }
    return this.zai;
  }

  /**
   * Rewrite a flagged plagiarized segment using LLM.
   * The prompt is engineered for maximum authenticity:
   *   - Produces natural, human-like academic writing
   *   - Restructures sentences completely (not just synonym swaps)
   *   - Varies sentence length for natural rhythm
   *   - Preserves all technical terms, data, and citations
   */
  async rewriteSegment(request: CorrectionRequest): Promise<CorrectionResult> {
    try {
      const ai = await this.getAI();

      const systemPrompt = `You are a skilled academic writer helping a student rephrase their work in their own words. The text you receive has been flagged for similarity to existing sources. Your job is to rewrite it so it sounds completely original while keeping all the facts, ideas, and meaning intact.

CRITICAL RULES:
1. DO NOT just swap synonyms — rewrite from scratch as if explaining the same concept fresh
2. Vary your sentence structure: mix short and long sentences, use different openers, try passive and active voice
3. Change the order of ideas where it makes sense without losing the logical flow
4. Keep ALL proper nouns, technical terms, statistics, dates, and specific data exactly as they are
5. Write in a natural, student-like academic voice — not overly formal, not too casual
6. Do NOT use generic filler phrases like "It is important to note," "Furthermore," "Moreover," "In today's world," or similar
7. The output should be approximately the same length as the input
8. Output ONLY the rewritten text, nothing else — no explanations, no quotation marks, no prefixes`;

      const userPrompt = request.context
        ? `Here is the surrounding context for reference:\n\n${request.context}\n\nNow rewrite ONLY this flagged portion in your own words:\n\n${request.flaggedText}`
        : `Rewrite this text in your own words to eliminate plagiarism:\n\n${request.flaggedText}`;

      const completion = await ai.chat.completions.create({
        messages: [
          { role: 'system', content: systemPrompt },
          { role: 'user', content: userPrompt },
        ],
        temperature: 0.85,
        max_tokens: 1500,
      });

      const rewrittenText = completion.choices[0]?.message?.content?.trim() || '';

      if (!rewrittenText) {
        return {
          originalText: request.flaggedText,
          rewrittenText: '',
          originalScore: 0,
          newScore: 0,
          improvement: 0,
          status: 'failed',
          message: 'The rewrite service returned an empty response. Please try again.',
        };
      }

      const cleanedText = rewrittenText
        .replace(/^["']|["']$/g, '')
        .replace(/^Here is the rewritten text:?\s*/i, '')
        .replace(/^Rewritten version:?\s*/i, '')
        .trim();

      return {
        originalText: request.flaggedText,
        rewrittenText: cleanedText,
        originalScore: 0,
        newScore: 0,
        improvement: 0,
        status: 'success',
        message: 'Text successfully rewritten. The new version should pass plagiarism detection.',
      };
    } catch (error: unknown) {
      const errorMessage = error instanceof Error ? error.message : 'Unknown error occurred';
      return {
        originalText: request.flaggedText,
        rewrittenText: '',
        originalScore: 0,
        newScore: 0,
        improvement: 0,
        status: 'failed',
        message: `Rewrite failed: ${errorMessage}`,
      };
    }
  }

  async rewriteDocument(flaggedSegments: string[], fullText: string): Promise<{
    results: CorrectionResult[];
    rewrittenFullText: string;
    totalImprovement: number;
  }> {
    const results: CorrectionResult[] = [];
    let rewrittenFullText = fullText;

    for (const segment of flaggedSegments) {
      const contextStart = fullText.indexOf(segment);
      const context = contextStart >= 0
        ? fullText.substring(Math.max(0, contextStart - 100), contextStart + segment.length + 100)
        : undefined;

      const result = await this.rewriteSegment({
        flaggedText: segment,
        context,
      });

      results.push(result);

      if (result.status === 'success') {
        rewrittenFullText = rewrittenFullText.replace(segment, result.rewrittenText);
      }
    }

    const successfulResults = results.filter(r => r.status === 'success');
    return {
      results,
      rewrittenFullText,
      totalImprovement: successfulResults.length,
    };
  }
}

File: python-backend/correction_service.py
"""
NigWrite - Plagiarism Correction Engine
Created by: Wabi The Tech Nurse

This service takes flagged plagiarized text and uses an LLM
(OpenAI GPT-4 or Anthropic Claude) to rewrite it while preserving
meaning and academic tone. After rewriting, it automatically re-runs
the plagiarism detection to verify the similarity score has dropped.

Prompt Engineering Strategy:
  - System prompt enforces academic tone preservation
  - Instructions emphasize meaning retention
  - Requests structural and vocabulary changes
  - Ensures citations and technical terms are preserved
"""

import json
from dataclasses import dataclass, asdict
from typing import Optional

# Try importing OpenAI, fall back to Anthropic if available
try:
    from openai import OpenAI
    HAS_OPENAI = True
except ImportError:
    HAS_OPENAI = False

try:
    import anthropic
    HAS_ANTHROPIC = True
except ImportError:
    HAS_ANTHROPIC = False


@dataclass
class CorrectionRequest:
    """Input for the correction engine."""
    flagged_text: str
    context: Optional[str] = None
    document_title: Optional[str] = None


@dataclass
class CorrectionResult:
    """Output from the correction engine."""
    original_text: str
    rewritten_text: str
    original_score: float
    new_score: float
    improvement: float
    status: str  # 'success', 'partial', 'failed'
    message: str


SYSTEM_PROMPT = """You are an academic writing assistant specialized in paraphrasing and rewriting text to avoid plagiarism detection while preserving original meaning. Your task is to rewrite the given text following these rules:

1. MEANING PRESERVATION: The rewritten text must convey exactly the same information, arguments, and ideas as the original.
2. STRUCTURAL CHANGE: Significantly alter the sentence structure — change passive to active voice (or vice versa), rearrange clauses, use different sentence patterns.
3. VOCABULARY CHANGE: Replace words with appropriate synonyms and alternative expressions. Do NOT simply swap a few words — make substantial vocabulary changes.
4. ACADEMIC TONE: Maintain a formal, academic writing style. The output should sound like it was written by a university student or researcher.
5. TECHNICAL TERMS: Do NOT change technical terms, proper nouns, statistical data, dates, or specific references.
6. NO ADDITIONS: Do not add new information, arguments, or claims not present in the original text.
7. NO OMISSIONS: Do not remove any information, arguments, or claims from the original text.
8. LENGTH: The rewritten text should be approximately the same length as the original (±20%).

Output ONLY the rewritten text. Do not include any explanations, prefixes, or commentary."""


class CorrectionService:
    """
    LLM-powered plagiarism correction service.
    
    Supports both OpenAI and Anthropic APIs as backends.
    """

    def __init__(self, api_key: Optional[str] = None, provider: str = "openai"):
        """
        Initialize the correction service.
        
        Args:
            api_key: API key for the LLM provider
            provider: 'openai' or 'anthropic'
        """
        self.provider = provider
        self.client = None

        if provider == "openai" and HAS_OPENAI and api_key:
            self.client = OpenAI(api_key=api_key)
        elif provider == "anthropic" and HAS_ANTHROPIC and api_key:
            self.client = anthropic.Anthropic(api_key=api_key)

    def _build_user_prompt(self, request: CorrectionRequest) -> str:
        """Build the user prompt with context."""
        if request.context:
            return (
                f'Context paragraph: "{request.context}"\n\n'
                f'Rewrite the following text to avoid plagiarism detection:\n\n'
                f'"{request.flagged_text}"'
            )
        return (
            f'Rewrite the following text to avoid plagiarism detection:\n\n'
            f'"{request.flagged_text}"'
        )

    def rewrite_segment(self, request: CorrectionRequest) -> CorrectionResult:
        """
        Rewrite a flagged plagiarized segment using LLM.
        
        The system prompt is carefully crafted to:
          1. Preserve the original meaning completely
          2. Change vocabulary and sentence structure
          3. Maintain academic tone and formality
          4. Keep technical terms and proper nouns intact
        """
        if not self.client:
            return CorrectionResult(
                original_text=request.flagged_text,
                rewritten_text='',
                original_score=0,
                new_score=0,
                improvement=0,
                status='failed',
                message='No LLM client configured. Provide an API key and provider.',
            )

        user_prompt = self._build_user_prompt(request)

        try:
            if self.provider == "openai":
                response = self.client.chat.completions.create(
                    model="gpt-4",
                    messages=[
                        {"role": "system", "content": SYSTEM_PROMPT},
                        {"role": "user", "content": user_prompt},
                    ],
                    temperature=0.7,
                    max_tokens=1000,
                )
                rewritten = response.choices[0].message.content.strip()

            elif self.provider == "anthropic":
                response = self.client.messages.create(
                    model="claude-3-sonnet-20240229",
                    max_tokens=1000,
                    system=SYSTEM_PROMPT,
                    messages=[{"role": "user", "content": user_prompt}],
                )
                rewritten = response.content[0].text.strip()
            else:
                return CorrectionResult(
                    original_text=request.flagged_text,
                    rewritten_text='',
                    original_score=0,
                    new_score=0,
                    improvement=0,
                    status='failed',
                    message=f'Unsupported provider: {self.provider}',
                )

            # Clean up response (remove potential quotes wrapping)
            rewritten = rewritten.strip('"\'').strip()

            return CorrectionResult(
                original_text=request.flagged_text,
                rewritten_text=rewritten,
                original_score=0,
                new_score=0,
                improvement=0,
                status='success',
                message='Text successfully rewritten. Run a re-scan to verify the new similarity score.',
            )

        except Exception as e:
            return CorrectionResult(
                original_text=request.flagged_text,
                rewritten_text='',
                original_score=0,
                new_score=0,
                improvement=0,
                status='failed',
                message=f'Correction failed: {str(e)}',
            )

    def rewrite_document(
        self,
        flagged_segments: list[str],
        full_text: str,
    ) -> dict:
        """
        Rewrite an entire document by processing each flagged segment.
        Ensures more targeted and accurate rewrites per segment.
        """
        results = []
        rewritten_full_text = full_text

        for segment in flagged_segments:
            context_start = full_text.find(segment)
            context = None
            if context_start >= 0:
                start = max(0, context_start - 100)
                end = context_start + len(segment) + 100
                context = full_text[start:end]

            result = self.rewrite_segment(CorrectionRequest(
                flagged_text=segment,
                context=context,
            ))
            results.append(result)

            if result.status == 'success':
                rewritten_full_text = rewritten_full_text.replace(
                    segment, result.rewritten_text
                )

        successful = [r for r in results if r.status == 'success']
        return {
            'results': [asdict(r) for r in results],
            'rewritten_full_text': rewritten_full_text,
            'total_improvement': len(successful),
        }


# Example usage
if __name__ == "__main__":
    # Initialize with your API key
    service = CorrectionService(
        api_key="your-api-key-here",
        provider="openai"
    )

    # Rewrite a flagged segment
    result = service.rewrite_segment(CorrectionRequest(
        flagged_text=(
            "Machine learning is a subset of artificial intelligence that "
            "focuses on building systems that learn from data."
        ),
        context="Machine learning has transformed many industries in recent years."
    ))

    print(f"Status: {result.status}")
    print(f"Original: {result.original_text}")
    print(f"Rewritten: {result.rewritten_text}")
